bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDevice.java

```diff
@@ -43,10 +47,15 @@ public HmDevice(String address, HmInterface hmInterface, String type, String gat
             String firmware) {
         this.address = address;
         this.hmInterface = hmInterface;
-        this.type = type;
+        this.firmware = firmware;
+        if ("HM-ES-TX-WM".equals(type) && Float.valueOf(firmware) > 2.0) {
+            logger.debug("Found HM-ES-TX-WM with firmware version > 2.0, creating virtual type");
+            this.type = type + "2";
+        } else {
+            this.type = type;
+        }
```

Comment:
@MHerbst Having stumbled upon this only now, I'm curious: What's the reason for this logic? Datapoint caching already factors in firmware versions, so at which point is there an issue when not doing this adaption?

(Generally I think adding workarounds like this might impair maintainability in the future, so I think we should strive for a more general solution here)

Reply:
@maniac103 This device is very special, because the available data points are completely different for the different firmware versions (IMO a very bad design vom EQ3). The HM binding only supports the same firmware version for all devices of the same type.
If you need to have two of these devices with different firmware, then it was not possible. Therefore the decision to handle  the combination of device and firmware version as different devices (= thing types).

> (Generally I think adding workarounds like this might impair maintainability in the future, so I think we should strive for a more general solution here)

I agree with you. There are some more "workarounds" in the binding for some other devices. I already thought about an additional configuration file, that a) make maintenance easier and b) would allow a user to influence the creation of channels and configuration options. Some devices have dozens of data points that are not needed with OH. But this would mean some larger refactorings in the binding.

Reply:
> the combination of device and firmware version as different devices (= thing types).

Ah, that's the key: Thing type UID generation doesn't factor in the firmware version. I wonder whether exploring changing the thing type UID to include the firmware version would be an option? openhab-core seems to provide hooks for that (see [here](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/binding/thinghandlercallback#migrateThingType(org.openhab.core.thing.Thing,org.openhab.core.thing.ThingTypeUID,org.openhab.core.config.core.Configuration)) and [here](https://next.openhab.org/javadoc/latest/org/openhab/core/thing/binding/basethinghandler#changeThingType(org.openhab.core.thing.ThingTypeUID,org.openhab.core.config.core.Configuration)) ... not sure how much work that would involve.

Reply:
I don't know whether it could be a problem if the firmware changes a bit later. Don't know whether there are any bindings that include the firmware version into the UID.